pkg/remote/kyma_synchronization_context.go

```diff
@@ -174,11 +225,24 @@ func (c *KymaSynchronizationContext) CreateOrFetchRemoteKyma(
 
 	if meta.IsNoMatchError(err) {
 		recorder.Event(kyma, "Normal", err.Error(), "CRDs are missing in SKR and will be installed")
+	}
 
-		if err := c.CreateOrUpdateCRD(ctx, v1beta2.KymaKind.Plural()); err != nil {
+	if meta.IsNoMatchError(err) || err == nil {
+		var kcpCrd, skrCrd *v1extensions.CustomResourceDefinition
+		if kcpCrd, skrCrd, err = CreateOrUpdateCRD(
+			ctx, v1beta2.KymaKind.Plural(), kyma, c.RuntimeClient, c.ControlPlaneClient); err != nil {
 			return nil, err
 		}
 
+		if !ContainsLatestCRDGeneration(kyma.Annotations[v1beta2.KcpKymaCRDGenerationAnnotation],
+			strconv.FormatInt(kcpCrd.Generation, 10)) ||
+			!ContainsLatestCRDGeneration(kyma.Annotations[v1beta2.SkrKymaCRDGenerationAnnotation],
+				strconv.FormatInt(skrCrd.Generation, 10)) {
+			UpdateKymaAnnotations(kyma, kcpCrd, skrCrd)
+			if err = c.ControlPlaneClient.Update(ctx, kyma); err != nil {
```

Comment:
don't do update kyma in the middle of reconciler loop, always update kyma in the end of reconciliation, mainly two reason:

1. save request to api server, each update is a remote http call, it's expensive regarding performance. 
2. after you update, the kyma in memory is different than the actual kyma in cluster, if you use this in memory kyma update later, it will failed, because the optimistic lock designed by apiserver to prevent old resource get write into etcd.